File: src/com/aslan/module/random/engines/DARandomEngine.js
import RandomEngine from "../RandomEngine";
import {S1, S2} from "../box";

export default class DARandomEngine extends RandomEngine {
	static DEFAULT_BUF_SIZE = 128;
	N = 0;
	H = 0;
	R = 0;
	D = null;	//Seed array
	B = null;	//Buffer array
	seedIndex = 0;
	index = 0;

	constructor() {
		super();
		this.init(DARandomEngine.DEFAULT_BUF_SIZE);
	}

	init(size) {
		this.N = size;
		this.B = new Uint8Array(size);
		this.D = new Uint8Array(size);
		this.H = this.N >> 1;
		this.R = Math.log2(size);
	}

	seed(seed, offset, length) {
		let min = 0, remainder = Math.min(length, this.N), n = this.N - 1;
		while (remainder > 0) {
			min = Math.min(remainder, this.N - this.seedIndex);
			this.D.set(new Uint8Array(seed.buffer, offset, min), this.seedIndex);
			offset += min;
			remainder -= min;
			this.seedIndex = (this.seedIndex + min) & n;
		}
		this.update();
	}

	next(accept, offset, length) {
		let min = 0, remainder = length;
		while (remainder > 0) {
			min = Math.min(remainder, this.N);
			accept.set(new Uint8Array(this.B.buffer, 0, min), offset);
			offset += min;
			remainder -= min;
			this.update();
		}
		return length;
	}

	rand() {
		let value = this.B[this.index++];
		if (this.index >= this.N) {
			this.index = 0;
			this.update();
		}
		return value;
	}

	get needSeed() {
		return this.N;
	}


	update() {
		let R = this.R, H = this.H, N = this.N;
		let B = this.B, D = this.D;
		let x = 0, r = R - 1, h = H - 1;
		for (let i = 0; i < r; i++) {
			for (let y = H; y < N;) {
				B[y] = S1[(B[y] ^ D[y])] ^ B[x];
				B[x] = S2[(B[x] ^ D[x])] ^ B[y];
				x = (++x) & h, y++;
				B[y] = S1[(B[y] ^ D[y])] ^ B[x];
				B[x] = S2[(B[x] ^ D[x])] ^ B[y];
				x = (++x) & h, y++;
				B[y] = S1[(B[y] ^ D[y])] ^ B[x];
				B[x] = S2[(B[x] ^ D[x])] ^ B[y];
				x = (++x) & h, y++;
				B[y] = S1[(B[y] ^ D[y])] ^ B[x];
				B[x] = S2[(B[x] ^ D[x])] ^ B[y];
				x = (++x) & h, y++;
			}
			x = H >> i;
		}
	}
}

File: src/com/aslan/module/random/engines/DARandomEngine.java
package com.aslan.module.random.engines;

import com.aslan.module.random.RandomEngine;
import com.aslan.module.utils.Utils;

import static com.aslan.module.random.Box.*;

public class DARandomEngine implements RandomEngine {
    static int DEFAULT_BUF_SIZE = 128;
    int N = 0;
    int H = 0;
    int R = 0;
    byte[] D = null;    //Seed array
    byte[] B = null;    //Buffer array
    int seedIndex = 0;
    int index = 0;

    public DARandomEngine() {
        init(DARandomEngine.DEFAULT_BUF_SIZE);
    }

    void init(int size) {
        N = size;
        B = new byte[size];
        D = new byte[size];
        H = N >> 1;
        R = (int) Utils.log2(size) + 1;
    }

    public void seed(byte[] seed, int offset, int length) {
        int min, remainder = Math.min(length, N), n = N - 1;
        while (remainder > 0) {
            min = Math.min(remainder, N - seedIndex);
            System.arraycopy(seed, offset, D, seedIndex, min);
            offset += min;
            remainder -= min;
            seedIndex = (seedIndex + min) & n;
        }
        update();
    }

    public int next(byte[] accept, int offset, int length) {
        int min = 0, remainder = length;
        while (remainder > 0) {
            min = Math.min(remainder, N);
            System.arraycopy(B, 0, accept, offset, min);
            offset += min;
            remainder -= min;
            update();
        }
        return length;
    }

    public byte rand() {
        byte value = B[index++];
        if (index >= N) {
            index = 0;
            update();
        }
        return value;
    }

    public int need() {
        return N;
    }


    protected void update() {
        int x = 0, r = R - 1, h = H - 1;
        for (int i = 0; i < r; i++) {
            for (int y = H; y < N; ) {
                B[y] = (byte) (S1[(B[y] ^ D[y]) & 0xFF] ^ B[x]);
                B[x] = (byte) (S2[(B[x] ^ D[x]) & 0xFF] ^ B[y]);
                x = (++x) & h;
                y++;
                B[y] = (byte) (S1[(B[y] ^ D[y]) & 0xFF] ^ B[x]);
                B[x] = (byte) (S2[(B[x] ^ D[x]) & 0xFF] ^ B[y]);
                x = (++x) & h;
                y++;
                B[y] = (byte) (S1[(B[y] ^ D[y]) & 0xFF] ^ B[x]);
                B[x] = (byte) (S2[(B[x] ^ D[x]) & 0xFF] ^ B[y]);
                x = (++x) & h;
                y++;
                B[y] = (byte) (S1[(B[y] ^ D[y]) & 0xFF] ^ B[x]);
                B[x] = (byte) (S2[(B[x] ^ D[x]) & 0xFF] ^ B[y]);
                x = (++x) & h;
                y++;

            }
            x = H >> i;
        }
    }
}